cpp/src/arrow/compute/kernels/vector_sort.cc

```diff
@@ -1909,6 +1909,110 @@ class SelectKUnstableMetaFunction : public MetaFunction {
+     "when ties appear in the input\n"
+     "\n"
+     "The handling of nulls, NaNs and tiebreakers can be changed in RankOptions."),
+    {"input"}, "RankOptions");
+
+class RankMetaFunction : public MetaFunction {
+ public:
+  RankMetaFunction()
+      : MetaFunction("rank", Arity::Unary(), &rank_doc, GetDefaultSortOptions()) {}
+
+  Result<Datum> ExecuteImpl(const std::vector<Datum>& args,
+                            const FunctionOptions* options, ExecContext* ctx) const {
+    const RankOptions& sort_options = static_cast<const RankOptions&>(*options);
+    switch (args[0].kind()) {
+      case Datum::ARRAY: {
+        return Rank(*args[0].make_array(), sort_options, ctx);
+      } break;
+      default:
+        break;
+    }
+    return Status::NotImplemented(
+        "Unsupported types for rank operation: "
+        "values=",
+        args[0].ToString());
+  }
+
+ private:
+  Result<Datum> Rank(const Array& array, const RankOptions& options,
+                     ExecContext* ctx) const {
+    ArraySortOptions array_options(options.order, options.null_placement);
+
+    ARROW_ASSIGN_OR_RAISE(auto sortIndices, CallFunction("array_sort_indices", {array},
+                                                         &array_options, ctx));
+
+    auto out_size = array.length();
+    ARROW_ASSIGN_OR_RAISE(auto rankings,
+                          MakeMutableUInt64Array(uint64(), out_size, ctx->memory_pool()));
+
+    auto* indices = sortIndices.make_array()->data()->GetValues<uint64_t>(1);
+    auto out_rankings = rankings->GetMutableValues<uint64_t>(1);
+    uint64_t rank = 0;
+    Datum prevValue, currValue;
+
+    if (options.tiebreaker == Tiebreaker::Dense) {
+      for (auto i = 0; i < out_size; i++) {
+        currValue = array.GetScalar(indices[i]).ValueOrDie();
+        if (i > 0 && currValue == prevValue) {
+        } else {
+          ++rank;
+        }
+        out_rankings[indices[i]] = rank;
+        prevValue = currValue;
+      }
+    } else if (options.tiebreaker == Tiebreaker::First) {
+      for (auto i = 0; i < out_size; i++) {
+        rank = i + 1;
+        out_rankings[indices[i]] = rank;
+      }
+    } else if (options.tiebreaker == Tiebreaker::Lowest) {
+      for (auto i = 0; i < out_size; i++) {
+        currValue = array.GetScalar(indices[i]).ValueOrDie();
+        if (i > 0 && currValue == prevValue) {
+        } else {
+          rank = i + 1;
+        }
+        out_rankings[indices[i]] = rank;
+        prevValue = currValue;
+      }
+    } else if (options.tiebreaker == Tiebreaker::Highest) {
+      auto currentTieCount = 0;
+      for (auto i = 0; i < out_size; i++) {
+        currValue = array.GetScalar(indices[i]).ValueOrDie();
+        if (i > 0 && currValue == prevValue) {
+          currentTieCount++;
+        } else {
+          currentTieCount = 0;
+        }
+        rank = i + 1;
+
+        // This can be inefficient when dealing many tied values
```

Comment:
Why not simply loop on the array in reverse order?

Reply:
That might work here but I *think* gets a little dicey if we later implement null handling that doesn't rank null values at all; in such a case I think we'd have to add an extra compute step to count the number of nulls to know which number to start at before looping in reverse. Definitely an option though.

My other idea was that we could look ahead to the next value instead of the previous one and only commit values when the next value changes.

Reply:
Think this should be good now. Lastly looking at implementing the visitor pattern to wrap up the inefficient array access you mentioned in another comment then ready for another look